cni/network/network.go

```diff
@@ -487,26 +489,35 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 		}
 	}
 
+	switch nwCfg.Ipam.Type {
+	case azureCNS:
+		plugin.ipamInvoker, err = NewCNSInvoker(k8sPodName, k8sNamespace)
+		if err != nil {
+			log.Printf("[cni-net] Creating network %v.", networkId)
+			return err
+		}
+	default:
+		plugin.ipamInvoker = NewAzureInvoker(plugin)
+	}
+
 	if nwInfoErr != nil {
 		// Network does not exist.
 		log.Printf("[cni-net] Creating network %v.", networkId)
 
 		if !nwCfg.MultiTenancy {
-			result, resultV6, err = plugin.invokeIpamAdd(*nwCfg, nwInfo, true)
+			result, resultV6, err = plugin.ipamInvoker.Add(args, nwCfg, nwInfo, true)
 			if err != nil {
 				return err
 			}
 
 			defer func() {
 				if err != nil {
-					plugin.invokeIpamDel(result, nwCfg.Ipam.Type, *nwCfg, true)
-					plugin.invokeIpamDel(resultV6, ipamV6, *nwCfg, true)
+					plugin.ipamInvoker.Delete(result, resultV6, nwCfg, true)
```

Comment:
Why does defer in Add() use invokeIpamDel, but then in the actual Delete call it calls delegateDel